### PR TITLE
feat: Kds 656 implement smarts to avoid really long select buttons

### PR DESCRIPTION
**Pull request by NatalieLoyzaga**
## Why

- truncate label options based on character limit rather than options length.

## What

- Add optional prop to the Trigger Button to accept char limit
- Update truncation function to recursively check against char limit
- Add story to show the new functionality

### CODE REVIEW

File: packages/select/src/FilterMultiSelect/utils/getTruncatedLabel/getTruncatedLabel.ts [hunk 1]
```diff
@@ -1,13 +1,22 @@
...
+  labelsTruncatedCount = 0
+): string => {
+  const moreIndicator =
+    labelsTruncatedCount > 0 ? ` + ${labelsTruncatedCount} more` : ""
+  if (labels.length === 1) {
+    return `${labels.join(CONNECTOR)}${moreIndicator}`
+  }
+  if (labels.join("").length < limit) {
+    return `${labels.join(CONNECTOR)}${moreIndicator}`
   }
```
Comment (HeartSquared) [inline packages/select/src/FilterMultiSelect/utils/getTruncatedLabel/getTruncatedLabel.ts:15]:
```suggestion
  if (labels.length === 1 || labels.join("").length < limit) {
    const moreIndicator =
      labelsTruncatedCount > 0 ? ` + ${labelsTruncatedCount} more` : ""
    return `${labels.join(CONNECTOR)}${moreIndicator}`
  }
```

File: packages/select/src/FilterMultiSelect/utils/getTruncatedLabel/getTruncatedLabel.ts [hunk 2]
```diff
@@ -1,13 +1,22 @@
...
+  labels: string[],
+  limit: number,
+  labelsTruncatedCount = 0
+): string => {
+  const moreIndicator =
+    labelsTruncatedCount > 0 ? ` + ${labelsTruncatedCount} more` : ""
+  if (labels.length === 1) {
+    return `${labels.join(CONNECTOR)}${moreIndicator}`
+  }
+  if (labels.join("").length < limit) {
```
Comment (HeartSquared) [inline packages/select/src/FilterMultiSelect/utils/getTruncatedLabel/getTruncatedLabel.ts:13]: We definitely don't want to count the `, ` as part of the length?

File: packages/select/src/FilterMultiSelect/components/Trigger/FilterTriggerButton/FilterTriggerButton.tsx [hunk 3]
```diff
@@ -7,12 +7,14 @@ export type FilterTriggerButtonProps = {
   label: string
   selectedOptionLabels: string[]
   classNameOverride?: string // TODO: migrate it to use OverrideClassName<T> and omit the props controlled by React-Aria
+  characterLimit?: number
```
Comment (HeartSquared) [inline packages/select/src/FilterMultiSelect/components/Trigger/FilterTriggerButton/FilterTriggerButton.tsx:10]: Hmm...given this prop name, I would expect it to cap the label length, however the functionality would actually still display a full length long string if it were the first option to display.

Perhaps clarify with @jmitsios if we want to actually truncate the first label if it exceeds the character limit? If not, we should try and think of a better name. 

It could benefit from a description regardless too :)

File: packages/select/docs/FilterMultiSelect.stories.tsx [hunk 4]
```diff
@@ -69,6 +70,63 @@ export const DefaultKaizenSiteDemo = args => {
 
 DefaultKaizenSiteDemo.storyName = "Default (Kaizen Site Demo)"
 
+export const TruncatedLabels = args => {
```
Comment (HeartSquared) [inline packages/select/docs/FilterMultiSelect.stories.tsx:73]: Do you want this story to have controls? If so, you would want to spread the `args` into the story somewhere. Otherwise, disable the controls and change `args` to `()`.

Bonus points if you type the story too with either `Story` or `ComponentStory` :)

File: packages/select/docs/FilterMultiSelect.stories.tsx [hunk 5]
```diff
@@ -69,6 +70,63 @@ export const DefaultKaizenSiteDemo = args => {
...
+  const [selectedKeys, setSelectedKeys] = useState<Selection>(
+    new Set(["id-fe"])
+  )
+  const [characterLimit, setCharacterLimit] = useState<number>(100)
+
+  const handleSelectionChange = (keys: Selection) => {
+    keys && setSelectedKeys(keys)
+  }
```
Comment (HeartSquared) [inline packages/select/docs/FilterMultiSelect.stories.tsx:81]:
```suggestion
  const handleSelectionChange = (keys: Selection) => setSelectedKeys(keys)
```

Based on your params, `keys` will always be defined

File: packages/select/docs/FilterMultiSelect.stories.tsx [hunk 6]
```diff
@@ -69,6 +70,63 @@ export const DefaultKaizenSiteDemo = args => {
...
+
+  const handleCharacterLimitChange = e => {
```
Comment (HeartSquared) [inline packages/select/docs/FilterMultiSelect.stories.tsx:83]: Type please :)

File: packages/select/docs/FilterMultiSelect.stories.tsx [hunk 7]
```diff
@@ -69,6 +70,63 @@ export const DefaultKaizenSiteDemo = args => {
...
+      <div style={{ marginBottom: "3rem" }}>
+        <Label
+          labelText="Update character limit: "
+          style={{ marginRight: "1rem" }}
+        />
+        <input
+          type="number"
+          onChange={handleCharacterLimitChange}
+          value={characterLimit}
+        ></input>
```
Comment (HeartSquared) [inline packages/select/docs/FilterMultiSelect.stories.tsx:98]:
```suggestion
        />
```

File: packages/select/src/FilterMultiSelect/utils/getTruncatedLabel/getTruncatedLabel.ts [hunk 8]
```diff
@@ -1,13 +1,22 @@
...
   }
 
-  return `${[...labels].splice(0, LABELS_TRUNCATED_LIMIT).join(CONNECTOR)} +${
-    labels.length - LABELS_TRUNCATED_LIMIT
-  } more`
+  const newLabels = labels.slice(0, labels.length - 1)
+  return truncateByCharacterLimit(newLabels, limit, labelsTruncatedCount + 1)
 }
+
+export const getTruncatedLabels = (labels: string[], limit: number): string =>
```
[the remark this answers is not in the document]
Comment (JakePitman) [reply packages/select/src/FilterMultiSelect/utils/getTruncatedLabel/getTruncatedLabel.ts:21]: Yeah - added this to remove the possibility of setting the `labelsTruncatedProp` manually from the outside. It's purely there for recursion. So `getTruncatedLabels` is the same thing, minus the third param. I _did_ forget to stop exporting the recursive function though, which defeats the whole purpose... I'll fix that

File: packages/select/src/FilterMultiSelect/components/Trigger/FilterTriggerButton/FilterTriggerButton.tsx (hunk 3 above)
Comment (NatalieLoyzaga) [reply packages/select/src/FilterMultiSelect/components/Trigger/FilterTriggerButton/FilterTriggerButton.tsx:10]: We discussed this in the kickoff and the first label regardless of its char length should always display. We also discussed there isn't an obvious name to convey that obscurity. I will add a description which will defs help.

File: packages/select/src/FilterMultiSelect/utils/getTruncatedLabel/getTruncatedLabel.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (HeartSquared) [reply packages/select/src/FilterMultiSelect/utils/getTruncatedLabel/getTruncatedLabel.ts:18]: If the third param is optional, then it should be to use anyway :)

File: packages/select/src/FilterMultiSelect/components/Trigger/FilterTriggerButton/FilterTriggerButton.tsx (hunk 3 above)
Comment (HeartSquared) [reply packages/select/src/FilterMultiSelect/components/Trigger/FilterTriggerButton/FilterTriggerButton.tsx:14]: Hmm yeah that's hard 🤔 I can't think of anything better currently either. A description will do for now :)

File: packages/select/docs/FilterMultiSelect.stories.tsx [hunk 9]
```diff
@@ -29,25 +30,25 @@ export default {
...
-      label="Engineer"
+      {...args}
       onSelectionChange={handleSelectionChange}
       selectedKeys={selectedKeys}
       items={items}
       trigger={() => (
         <FilterMultiSelect.TriggerButton
           selectedOptionLabels={getSelectedOptionLabels(selectedKeys, items)}
-          label="Engineer"
+          {...args}
```
Comment (HeartSquared) [inline packages/select/docs/FilterMultiSelect.stories.tsx:51]: Hmm... This probably doesn't make sense, because not all props for `FilterMultiSelect` would apply to `TriggerButton`.

This could probably work though?

```suggestion
          label={args.label}
```

File: packages/select/src/FilterMultiSelect/components/Trigger/FilterTriggerButton/FilterTriggerButton.tsx (hunk 3 above)
Comment (dougmacknz) [reply packages/select/src/FilterMultiSelect/components/Trigger/FilterTriggerButton/FilterTriggerButton.tsx:14]: > The first label regardless of its char length should always display

Definitely agree but now I feel like we should probably account for how long the label is in this calculation (if that's not happening already?) 

Is this too wordy: `labelCharacterLimitBeforeTruncate`?

Comment (HeartSquared) [reply packages/select/src/FilterMultiSelect/components/Trigger/FilterTriggerButton/FilterTriggerButton.tsx:14]: If it adds clarity, there's nothing wrong with a long variable imo. Shippit

**Review by HeartSquared — APPROVED** [review]
Code looks fine. Give some consideration to Doug's suggestion for the prop name though (since it would be a breaking change if you wanted to change it later).

**Review by JakePitman — APPROVED** [review]

**Review by JakePitman — APPROVED** [review]
